pkg/kubernetes/kubernetes.go

```diff
@@ -146,13 +152,92 @@ func NewInCluster(l *zap.SugaredLogger) (*Kubernetes, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// Create the cache for the cached read client and registering informers
+	syncPeriod := 1 * time.Hour
+	cacheOpts := ctrlcache.Options{
+		SyncPeriod: &syncPeriod,
+	}
+	cache, err := ctrlcache.New(client.Config(), cacheOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the client with cache, and default its options.
+	clientOpts := ctrlclient.Options{
+		Cache: &ctrlclient.CacheOptions{
+			Unstructured: false,
+			Reader:       cache,
+			// do not cache the following resources because in big K8S clusters
+			// it can lead to high memory usage
+			DisableFor: []ctrlclient.Object{
+				&corev1.ConfigMap{},
+				&corev1.Secret{},
+			},
+		},
+	}
+	clientWriter, err := ctrlclient.New(client.Config(), clientOpts)
+	if err != nil {
+		return nil, err
+	}
+
 	return &Kubernetes{
 		client:    client,
+		k8sCache:  cache,
+		k8sClient: clientWriter,
 		l:         l,
 		namespace: client.Namespace(),
 	}, nil
 }
 
+// StartCache starts the cache in background.
+func (k *Kubernetes) StartCache(ctx context.Context) error {
+	if err := k.initIndexers(ctx); err != nil {
+		return errors.Join(err, errors.New("failed to init field indexers for cache"))
+	}
+
+	// Start the cache in a separate goroutine, since it is a blocking call.
+	go func() {
+		if err := k.k8sCache.Start(ctx); err != nil {
+			k.l.Error("failed to start cache", zap.Error(err))
+		}
+	}()
+	return nil
+}
+
+// initIndexers initializes the field indexers for the DatabaseCluster.
+func (k *Kubernetes) initIndexers(ctx context.Context) error {
+	if err := k.k8sCache.IndexField(
+		ctx, &everestv1alpha1.DatabaseCluster{}, ".spec.backup.schedules.backupStorageName",
```

Comment:
this is an example of indexing field in `.spec` for our DatabaseCluster CR. This index is later used for DBs lookup by field